include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h

```diff
@@ -507,23 +507,18 @@ namespace llvm {
 template <typename To, typename From>
 struct CastInfo<
     To, From,
-    std::enable_if_t<
-        std::is_same_v<circt::firrtl::AnnoTarget, std::remove_const_t<From>> ||
-        std::is_base_of_v<circt::firrtl::AnnoTarget, From>>>
+    std::enable_if_t<std::is_base_of_v<::circt::firrtl::AnnoTarget, From>>>
     : NullableValueCastFailed<To>,
       DefaultDoCastIfPossible<To, From, CastInfo<To, From>> {
   static inline bool isPossible(From target) {
-    /// Return a constant true instead of a dynamic true when casting to self or
-    /// up the hierarchy.
-    if constexpr (std::is_base_of_v<To, From>) {
+    // Allow constant upcasting.  This also gets around the fact that AnnoTarget
+    // does not implement classof.
```

Comment:
👍 

Leaves `To` open to non-AnnoTarget's but I think it's good as-is.
What's annoying, kinda, is that isPossible I think by default is these things (base_of_v for statically resolving when self/upcast, `T::classof` otherwise) but we need to tell it how casts for these work since value-value.

Anyway! Looks great, thanks!

Reply:
Whatever you think is best, thank you very much.

As-is or two ones if you'd prefer :+1:. 🙏

Reply:
I tried a couple things but couldn't get it to work, so I just left it as is.  We can always make this better in a follow up PR.